src/wrapper-utils.js

```diff
@@ -47,3 +131,300 @@ export function wrapFunction (functionValue, realTarget) {
+    const defaultOptions = {
+        allowConstructorCall: false,
+        disallowConstructor: false,
+        constructorErrorMessage: 'Illegal constructor',
+        wrapToString: true
+    }
+
+    const fullOptions = {
+        interfaceDescriptorOptions: { writable: true, enumerable: false, configurable: true, value: ImplClass },
+        ...defaultOptions,
+        ...options
+    }
+
+    // In some cases we can get away without a full proxy, but in many cases below we need it.
+    // For example, we can't redefine `prototype` property on ES6 classes.
+    // Se we just always wrap the class to make the code more maintaibnable
+
+    /** @type {ProxyHandler<Function>} */
+    const proxyHandler = {}
+
+    // handle the case where the constructor is called without new
+    if (fullOptions.allowConstructorCall) {
+        // make the constructor function callable without new
+        proxyHandler.apply = function (target, thisArg, argumentsList) {
+            return Reflect.construct(target, argumentsList, target)
+        }
+    }
+
+    // make the constructor function throw when called without new
+    if (fullOptions.disallowConstructor) {
+        proxyHandler.construct = function () {
+            throw new TypeError(fullOptions.constructorErrorMessage)
+        }
+    }
+
+    if (fullOptions.wrapToString) {
+        // mask toString() on class methods. `ImplClass.prototype` is non-configurable: we can't override or proxy it, so we have to wrap each method individually
+        for (const [prop, descriptor] of objectEntries(getOwnPropertyDescriptors(ImplClass.prototype))) {
+            if (prop !== 'constructor' && descriptor.writable && typeof descriptor.value === 'function') {
+                ImplClass.prototype[prop] = new Proxy(descriptor.value, {
+                    get: toStringGetTrap(descriptor.value, `function ${prop}() { [native code] }`)
+                })
+            }
+        }
+
+        // wrap toString on the constructor function itself
+        Object.assign(proxyHandler, {
+            get: toStringGetTrap(ImplClass, `function ${interfaceName}() { [native code] }`)
+        })
+    }
+
+    // Note that instanceof should still work, since the `.prototype` object is proxied too:
+    // Interface() instanceof Interface === true
+    // ImplClass() instanceof Interface === true
+    const Interface = new Proxy(ImplClass, proxyHandler)
+
+    // Make sure that Interface().constructor === Interface (not ImplClass)
+    if (ImplClass.prototype?.constructor === ImplClass) {
+        /** @type {StrictDataDescriptor} */
+        // @ts-expect-error - As long as ImplClass is a normal class, it should have the prototype property
+        const descriptor = getOwnPropertyDescriptor(ImplClass.prototype, 'constructor')
+        if (descriptor.writable) {
+            ImplClass.prototype.constructor = Interface
+        }
+    }
+
+    // mark the class as a shimmed class
+    definePropertyFn(ImplClass, ddgShimMark, {
```

Comment:
Can/should this be wrapped with an integration check?

Reply:
neat idea!